src/Cryptol/TypeCheck/Solver/Numeric/Sampling/Preconstraints.hs

```diff
@@ -0,0 +1,261 @@
+            "The following type constraint is not supported: " ++ pretty prop
+      where
+        proc2 con ts =
+          toPExp `traverse` ts >>= \case
+            [e1, e2] -> do
+              let pprop = con e1 e2
+              pure $ addPProps [pprop] precons
+            _ -> undefined -- bad number of args
+        proc1 con ts =
+          toPExp `traverse` ts >>= \case
+            [e] -> pure $ addPProps [con e] precons
+            _ -> undefined -- bad number of args
+    toPExp :: Type -> SamplingM PExp
+    toPExp typ = do
+      case typ of
+        TCon tcon ts -> case tcon of
+          -- type constants
+          TC tc -> case tc of
+            TCNum n -> pure . PEConst $ toQ n
+            -- TODO: `TCInf -> ...` is possible but I need to use Q' instead of Q
+            _ -> unsupported
+            -- type functions
+          TF tf -> case tf of
+            TCAdd -> proc2 (PEOp2 PAdd) ts
+            TCSub -> proc2 (PEOp2 PSub) ts
+            TCMul -> proc2 (PEOp2 PMul) ts
+            TCDiv -> proc2 (PEOp2 PDiv) ts
+            TCMod -> proc2 (PEOp2 PMod) ts
+            TCExp -> proc2 (PEOp2 PPow) ts
+            _ -> unsupported
+            where
+              proc2 con [t1, t2] = con <$> toPExp t1 <*> toPExp t2
+              proc2 _ _ = error "impossible"
+          _ -> unsupported
+        TVar tv -> pure $ PETerm 1 (iTVar tv)
+        TUser _ _ prop -> toPExp prop
+        TNewtype _new _tys -> unsupported
+        _ -> unsupported
+      where
+        unsupported = throwError . SamplingError "Preconstraints.fromProps" $ 
+          "The following numeric type is not supported: " ++ pretty typ
+
+        iTVar :: TVar -> Var
+        iTVar = \case
+          TVFree {} -> undefined -- shouldn't be dealing with free vars here
+          TVBound tparam ->
+            maybe undefined Var (elemIndex tparam tps)
+
+{-
+- Check that all `a mod n` have `n` a constant
+- Check that all `m^n` have `m` and `n` constant
+- Check that all `n*a` has at most one of `n`, `a` a variable
+- Replace `a mod n` with `b` and add equality `b = c*n + a`, where
+  `n` is a constant.
+- Replace `a - b` with `a + (-b)`
+- Apply distributivity
+- Apply commutativity of addition to combine constant terms at the end a sum
+- Apply commutativity of addition to combine terms in a sum of products
+- Evaluate operations over constants
+-}
+normalizePreconstraints :: Preconstraints -> SamplingM Preconstraints
+normalizePreconstraints precons = do
+  ((preprops', preprops''), nVars) <-
+    flip runStateT (nVars precons) . runWriterT $
+      normPProp `traverse` preprops precons
+  pure
+    precons
+      { preprops = preprops' <> preprops'',
+        nVars = nVars
+      }
+  where
+    normPProp :: PProp -> WriterT [PProp] (StateT Int SamplingM) PProp
+    normPProp = \case
+      PPEqual pe1 pe2 -> PPEqual <$> normPExp pe1 <*> normPExp pe2
+      PPNeq _a _b -> do
+        error "not sure how to handle Neq"
```

Comment:
Can user input trigger this case? If not, panic, if so, let's find a better way to display this error.